realm_export: Add 'Export permissions' table.

prakhar1144 opened the pull request:
Issue: https://github.com/zulip/zulip/issues/31201

* First commit: Adds 'Export permissions' table.
* Second commit: Fixes live update issue in the `start export` modal.

**Screenshots and screen captures:**

<img width="491" alt="Screenshot 2024-10-16 at 8 31 54 PM" src="https://github.com/user-attachments/assets/eb07e603-82c5-4431-88a1-a9d6e8cf2598">

<img width="491" alt="Screenshot 2024-10-16 at 8 32 09 PM" src="https://github.com/user-attachments/assets/fbc2c9e6-2351-4946-8210-cfc87bd60e54">



<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

prakhar1144 commented:
Why I used `true` / `false` instead of `Allowed` / `Not allowed` as specified in the issue -- CZO Discussion

prakhar1144 commented:
Currently the PR live updates the `Export permissions` panel and `Exporting private data for {users_consented_for_export_count} users ({total_users_count} users total).` text in the following cases:

* New user added
* User removed
* User deactivated
* User reactivated

Currently, it doesn't live-update when a user changes their `allow_private_data_export` setting. To achieve this, we'll have to make a few changes.
* send the event to the user who changed their setting + admins
* API changes in https://zulip.com/api/get-events#user_settings-update to indicate that the `property=allow_private_data_export`  when changed and sent to admin is for some another user. -- This will help to only live-update tables and not the user-setting of admin. **Does this sound good ? -- I'll start an #api-design thread if this sounds good broadly**

timabbott commented:
That API change direction is plausible. Another option would be to have the client have a "refresh" button for this widget, if we don't want to deal with the live-update complexity. 

BTW looking at the screenshots, the tab name for "Export permissions" needs to not be abbreviated.

prakhar1144 commented:
> That API change direction is plausible. Another option would be to have the client have a "refresh" button for this widget, if we don't want to deal with the live-update complexity.

Moved forward with the API changes approach. Adding "refresh" button to that table as well as below dropdown text doesn't looks good.

> BTW looking at the screenshots, the tab name for "Export permissions" needs to not be abbreviated.

Fixed this. -- Screenshot updated.

prakhar1144 commented:
Updated the PR with the `realm_export_consent` event type approach.

It doesn't handle the [race which we discussed](https://chat.zulip.org/#narrow/channel/378-api-design/topic/automate.20private.20data.20export/near/1963793) in CZO today. But an initial review would be helpful.

prakhar1144 commented on web/src/settings_exports.ts:
```diff
@@ -203,16 +318,62 @@ function show_start_export_modal(): void {
+
+            // Apply queued_export_consents on top of the received response.
+            // This loop has O(n^2) time-complexity, but 'queued_export_consents'
+            // would be of a small length, making it effectively O(n).
```
Is this assumption acceptable ?

timabbott replied:
It's basically fine from a practical performance standpoint, but also you should just make the data structure a `map` from user IDs to their consent status, and then this code will get way way simpler and it'll also naturally eliminate the performance issue. Almost always, when the server sends a list to the client, the client's first thing to do will be to convert into a `map` from the primary key to the data, since while lists are simpler as a wire protocol, maps are cheaper to update.

prakhar1144 replied:
Updated code to use map.

prakhar1144 commented:
> Updated the PR with the `realm_export_consent` event type approach.
> 
> It doesn't handle the [race which we discussed](https://chat.zulip.org/#narrow/channel/378-api-design/topic/automate.20private.20data.20export/near/1963793) in CZO today. But an initial review would be helpful.

The PR now handles the race. 

Tested it manually by applying `time.sleep(10)` in the server code to delay sending response after the db query has been performed. During that delay, performed the following actions (separately):
* add new user
* deactivate
* reactivate
* toggle consent

 to confirm visually + by printing that the table is not in the outdated state. The events sent during that duration are applied on top of the response received.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -3884,6 +3884,36 @@ paths:
+                              additionalProperties: false
+                              description: |
+                                Event sent to administrators when a user changes their [Let administrators export my private data]
+                                (/help/export-your-organization#configure-whether-administrators-can-export-your-private-data)
```
Does this line-wrapping of a link work? Generally we should use a reference link in any case to keep the prose nice, and line-wrap a bit earlier.

timabbott replied:
Fixed this and made some other revisions in the first commit, which I'll merge via a separate PR to lock in the API feature level change.

timabbott commented on web/src/settings_exports.ts:
```diff
@@ -248,3 +409,67 @@ export function set_up(): void {
+    } else {
+        // user reactivated
+        export_consents.push(export_consent);
+    }
```
This logic also gets a lot cleaner if using a `Map`.

prakhar1144 replied:
Done

timabbott commented on web/src/settings_exports.ts:
```diff
@@ -248,3 +409,67 @@ export function set_up(): void {
+    // We store the export_consent data received via events to apply them on top of
+    // the received response.
+    if (export_consents === undefined) {
+        queued_export_consents.push(export_consent);
```
`queued_export_consents` would also be a lot cleaner as a `Map`, since then you can just walk this map, overwriting onto the `Map` that we got from the API response.

prakhar1144 replied:
> `queued_export_consents` would also be a lot cleaner as a Map

Kept it as an array because it's element is of type `ExportConsent | number`. We only need the key `user_id` when we want to remove the element.

We simply walk the array and perform the update or delete operation on the map. -- O(n)

timabbott commented:
Since the remaining changes are just on the UI side, I split out #32055 to merge the API changes. Thanks @prakhar1144!

prakhar1144 commented:
Update in the recent push:

* Addressed the review comment i.e. Updated the code to use `Map` to store the received response from the server and related changes.

timabbott commented on web/src/settings_exports.ts:
```diff
@@ -132,6 +139,124 @@ export function populate_exports_table(exports: RealmExport[]): void {
+        }
+    }
+    return export_consent_list;
+}
```
Should this function be merging the `export_consents` data with `people.ts` data to return a list that contains `false` for any user whose consent status we don't specifically know?

As discussed on another PR, I think that might be required to correctly show new user accounts as not having consented if we've not received an explicit data value for them, at least until the server sends an explicit event for every newly created user regardless of consent status.

timabbott commented on web/src/settings_exports.ts:
```diff
@@ -248,3 +405,44 @@ export function set_up(): void {
+    }
+
+    export_consents.delete(user_id);
+    redraw_export_consents_list();
```
This logic could probably be optimized to not redraw in the event that the settings panel is no longer open, since I expect it gets redrawn every time its opened?

(I think `meta.loaded` gets set to true the first time the panel is opened, even if the user navigates elsewhere after?)

timabbott commented:
Looks good enough to want to only see deltas from here, merged, but see the above couple notes about quirks that are probably worth fixing while this is fresh. (The remaining possible bugs seem unlikely to be encountered by any users before we can fix them, and will be easy to review as diffs on top of this). Thanks @prakhar1144!
